pkg/network/node/egressip.go

```diff
@@ -0,0 +1,302 @@
+package node
+
+import (
+	"fmt"
+	"net"
+	"sync"
+	"syscall"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/util/sets"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+
+	networkapi "github.com/openshift/origin/pkg/network/apis/network"
+	"github.com/openshift/origin/pkg/network/common"
+	networkclient "github.com/openshift/origin/pkg/network/generated/internalclientset"
+
+	"github.com/vishvananda/netlink"
+)
+
+type nodeEgress struct {
+	nodeIP    string
+	egressIPs sets.String
+}
+
+type namespaceEgress struct {
+	vnid uint32
+
+	// claimedIP is the egress IP it wants (NetNamespace.EgressIP[0]), or "" for none
+	claimedIP string
+	// assignedIP is an egress IP actually in use on nodeIP
+	assignedIP string
+	nodeIP     string
+}
+
+type egressIPWatcher struct {
+	sync.Mutex
+
+	localIP string
+	oc      *ovsController
+
+	networkClient networkclient.Interface
+	iptables      *NodeIPTables
+
+	// from HostSubnets
+	nodesByNodeIP   map[string]*nodeEgress
+	nodesByEgressIP map[string]*nodeEgress
+
+	// From NetNamespaces
+	namespacesByVNID     map[uint32]*namespaceEgress
```

Comment:
vnid may be duplicate here but considering most of the fields are also used by namespacesByEgressIP, I think this is a good compromise.